src/clinical/id-generator.ts

```diff
@@ -0,0 +1,293 @@
+		},
+	};
+
+	try {
+		const response = await axios.get(
+			// `http://localhost:9001/${req.programId}/${req.submitterId}/${req.submitterDonorId}/${req.testInterval}/${req.family_relative_id}/${req.comorbidityTypeCode}/${req.entityType}`,
+			config.getConfig().tokenUrl() +
+				`${req.programId}/${req.submitterId}/${req.submitterDonorId}/${req.testInterval}/${req.family_relative_id}/${req.comorbidityTypeCode}/${req.entityType}`,
+			headers,
+		);
+		console.log('getId response: ' + response.data.entityId + ' - ' + response.data.entityType);
+		return parseInt(response.data.entityId);
+	} catch (e) {
+		console.log(e);
+		throw Error('Error fetching ids');
+	}
+}
+
+export async function setEntityIdsForDonors(donors: Donor[]) {
+	const donorsWithIds = donors.map(async (donor) => {
+		return await setEntityIds(donor);
+	});
+	return await Promise.all(donorsWithIds);
+}
+
+export async function setEntityIds(donor: PartialDonor) {
+	// -- DONOR --
+	const donorId = await getId({
+		...request,
+		programId: donor.programId as string,
+		submitterId: donor.submitterId as string,
+		submitterDonorId: donor.submitterId as string,
+		entityType: ClinicalEntitySchemaNames.DONOR,
+	});
+	donor.donorId = donorId;
+	console.log(donor.donorId);
+
+	// -- SPECIMEN --
+	if (donor.specimens && donor.specimens.length > 0) {
+		for (const specimen of donor.specimens) {
+			const submitterId = !specimen.clinicalInfo.submitter_specimen_id
+				? specimen.submitterId
+				: specimen.clinicalInfo.submitter_specimen_id;
+			const id = await getId({
+				...request,
+				programId: donor.programId as string,
+				submitterId: submitterId as string,
+				entityType: ClinicalEntitySchemaNames.SPECIMEN,
+			});
+			specimen.specimenId = id;
+			console.log('specimen.specimenId: ' + specimen.specimenId);
+
+			// -- SAMPLE --
+			const samples = specimen.samples;
+			for (const sample of samples) {
+				const id = await getId({
+					...request,
+					programId: donor.programId as string,
+					submitterId: sample.submitterId as string,
+					entityType: ClinicalEntitySchemaNames.REGISTRATION,
+				});
+				sample.sampleId = id;
+				console.log('sample.sampleId: ' + sample.sampleId);
+			}
+		}
+	}
+
+	// -- BIOMARKER --
+	// submitter_specimen_id
+	if (donor.biomarker && donor.biomarker.length > 0) {
+		for (const bm of donor.biomarker) {
+			const submitterBiomarkerId =
+				bm.clinicalInfo.submitter_specimen_id?.toString() ??
+				bm.clinicalInfo.submitter_primary_diagnosis_id?.toString() ??
+				bm.clinicalInfo.submitter_follow_up_id?.toString() ??
+				bm.clinicalInfo.submitter_treatment_id?.toString() ??
+				'null';
+
+			const id = await getId({
+				...request,
+				programId: donor.programId as string,
+				submitterId: submitterBiomarkerId,
+				submitterDonorId: bm.clinicalInfo.submitter_donor_id as string,
+				testInterval: bm.clinicalInfo.test_interval as string,
+				entityType: ClinicalEntitySchemaNames.BIOMARKER,
+			});
```

Comment:
When grabbing many values from the same place you can use destructuring variable assignment to simplify things:

```
			const {
				submitter_donor_id,
				submitter_specimen_id,
				submitter_primary_diagnosis_id,
				submitter_follow_up_id,
				submitter_treatment_id,
				test_interval,
			} = bm.clinicalInfo;

			const submitterBiomarkerId =
				submitter_specimen_id?.toString() ??
				submitter_primary_diagnosis_id?.toString() ??
				submitter_follow_up_id?.toString() ??
				submitter_treatment_id?.toString() ??
				'null';

			const id = await getId({
				...request,
				programId: donor.programId as string,
				submitterId: submitterBiomarkerId,
				submitterDonorId: submitter_donor_id as string,
				testInterval: test_interval as string,
				entityType: ClinicalEntitySchemaNames.BIOMARKER,
			});
```

Reply:
Also generally prefer descriptive variable names -- `biomarker` vs `bm`

Reply:
I have changed bm to biomarker and for all other entities. But I'd prefer not to use the destructured variable assignment here for now